nsm-driver/nsm.c

```diff
@@ -458,6 +459,149 @@ static struct miscdevice nsm_driver_miscdevice = {
 	.mode	= 0666
 };
 
+
+static int nsm_rng_read(struct hwrng *rng, void *data, size_t max, bool wait) {
+	int rc = 0;
+	uint8_t *resp_ptr = NULL;
+	uint64_t resp_len = 0;
+	uint64_t rand_len = 0;
+
+	struct nsm_kernel_message message;
+
+	/*
+	 * 69                          # text(9)
+	 * 47657452616E646F6D          # "GetRandom"
+	 */
+	uint8_t request[] = { 0x69, 0x47, 0x65, 0x74, 0x52, 0x61, 0x6E, 0x64, 0x6F, 0x6D };
+
+	/*
+	 * A1                          # map(1)
+	 *     69                      # text(9) - Name of field
+	 *         47657452616E646F6D  # "GetRandom"
+	 * A1                          # map(1) - The field itself
+	 *     66                      # text(6)
+	 *         72616E646F6D        # "random"
+	 *	# The rest of the response should be a byte array
+	 */
+	uint8_t response[] = { 0xA1, 0x69, 0x47, 0x65, 0x74, 0x52, 0x61, 0x6E, 0x64,
+		0x6F, 0x6D, 0xA1, 0x66, 0x72, 0x61, 0x6E, 0x64, 0x6F, 0x6D };
+
+	if (wait == true) {
+		mutex_lock(&nsm_lock);
+	} else {
+		if (mutex_trylock(&nsm_lock) == 0)
+			return -EBUSY;
+	}
+
+	/* The kernel message structure must be cleared */
+	memset(&message, 0, sizeof(message));
+
+	/* Set request message */
+	message.request.iov_len = sizeof(request);
+	message.request.iov_base = kmalloc(message.request.iov_len, GFP_KERNEL);
+	if (message.request.iov_base == NULL)
+		goto out;
+	memcpy(message.request.iov_base, request, sizeof(request));
+
+	/* Allocate space for response */
+	message.response.iov_len = NSM_RESPONSE_MAX_SIZE;
+	message.response.iov_base = kmalloc(message.response.iov_len, GFP_KERNEL);
+	if (message.response.iov_base == NULL)
+		goto out;
```

Comment:
The `message.request.iov_base` will leak here or does `message_delete()` check pointer and free it inside? As well as other `goto out` after this?

Reply:
message_delete cleans it up completely.